File: CSW/JavaScript/220919_최소원_로봇시뮬레이션.js
var fs = require('fs');
var input = fs.readFileSync(`/dev/stdin`).toString().trim().split('\n');

//A 가로, B 세로
let [A , B] = input[0].split(' ').map(e=>parseInt(e));
let [N , M] = input[1].split(' ').map(e=>parseInt(e));

const board = Array.from({length:B}, ()=>Array(A).fill(0));
const robots = [];
const directions = [];
let state = 0;

for(let i=0; i<N; i++){
    let [x, y, dir] = input[2+i].trim().split(' ');
    robots.push([B-parseInt(y), parseInt(x)-1]);
    board[B-parseInt(y)][parseInt(x)-1] = i+1;
    directions[i] = dir;
}

const turnL = (dir) =>{
    if(dir === 'N'){
        return 'W';
    }else if(dir ==='W'){
        return 'S';
    }else if(dir === 'S'){
        return 'E';
    }else if(dir === 'E'){
        return 'N';
    }
}

const turnR = (dir) =>{
    if(dir === 'N'){
        return 'E';
    }else if(dir ==='E'){
        return 'S';
    }else if(dir === 'S'){
        return 'W';
    }else if(dir === 'W'){
        return 'N';
    }
}
const goStraight = (ry, rx, dir) =>{
    if(dir === 'N'){
        return [ry-1, rx];
    }else if(dir ==='W'){
        return [ry, rx-1];
    }else if(dir === 'S'){
        return [ry+1, rx];
    }else if(dir === 'E'){
        return [ry, rx+1];
    }

}


for(let i=0; i<M; i++){
    let [robot, order, num] = input[2+N+i].trim().split(' ');
    let [ry, rx] = robots[parseInt(robot)-1];
    let dir = directions[parseInt(robot)-1];
    let [dy, dx] = [ry, rx];

    for(let j=0; j<parseInt(num); j++){
        if(order === 'L'){
            dir = turnL(dir);
            directions[parseInt(robot)-1] = dir;
        }else if(order === 'R'){
            dir = turnR(dir);
            directions[parseInt(robot)-1] = dir;
        }else if(order === 'F'){
            [dy, dx] = goStraight(ry, rx, dir);
            if(dy<0 || dx <0 || dy>=B || dx>=A){
                state = 1;
                console.log(`Robot ` + robot +` crashes into the wall`);
                break;
            }
            if(board[dy][dx]){
                state = 1;
                console.log(`Robot ` + robot +` crashes into robot ` + board[dy][dx]);
                break;
            }
            board[ry][rx] = 0;
            board[dy][dx] = parseInt(robot);
            robots[parseInt(robot)-1] = [dy, dx];
            [ry, rx] = [dy, dx];
        }
    }
    if(state) break;
}
if(!state){
    console.log('OK');
}

File: CSW/C,C++/220919_최소원_로봇시뮬레이션.cpp
#include <iostream>
#include <vector>
using namespace std;

int A, B, N, M;
int board[100][100];
vector<pair<int, int>> robots;
char directions[100];
int dy, dx, state;

void turnL(int robot, char dir) {
	if (dir == 'N') {
		directions[robot - 1] = 'W';
	}
	else if (dir == 'W') {
		directions[robot - 1] = 'S';
	}
	else if (dir == 'S') {
		directions[robot - 1] = 'E';
	}
	else if (dir == 'E') {
		directions[robot - 1] = 'N';
	}
}

void turnR(int robot, char dir) {
	if (dir == 'N') {
		directions[robot - 1] = 'E';
	}
	else if (dir == 'E') {
		directions[robot - 1] = 'S';
	}
	else if (dir == 'S') {
		directions[robot - 1] = 'W';
	}
	else if (dir == 'W') {
		directions[robot - 1] = 'N';
	}
}

int goStraight(int ry, int rx, char dir) {
	if (dir == 'N') {
		ry = ry - 1;
	}
	else if (dir == 'W') {
		rx = rx - 1;
	}
	else if (dir == 'S') {
		ry = ry + 1;
	}
	else if (dir == 'E') {
		rx = rx + 1;
	}

	dy = ry;
	dx = rx;
	if (ry < 0 || rx < 0 || ry >= B || rx >= A) return -1;
	if (board[ry][rx]) return 0;
	return 1;
}

int main() {
	cin >> A >> B;
	cin >> N >> M;

	for (int i = 0; i < N; i++) {
		int x, y;
		char dir;
		cin >> x >> y >> dir;
		robots.push_back({ B-y,x-1 });
		directions[i] = dir;
		board[B - y][x - 1] = i + 1;
	}

	for (int i = 0; i < M; i++) {
		int robot, num;
		char order;
		cin >> robot >> order >> num;
		
		for (int j = 0; j < num; j++) {
			int ry = robots[robot - 1].first;
			int rx = robots[robot - 1].second;
			char dir = directions[robot - 1];
			if (order == 'L') {
				turnL(robot, dir);
			}
			else if (order == 'R') {
				turnR(robot, dir);
			}
			else if (order == 'F') {
				int result = goStraight(ry, rx, dir);
				if (result == 1) {
					board[ry][rx] = 0;
					board[dy][dx] = robot;
					robots[robot - 1].first = dy;
					robots[robot - 1].second = dx;
				}
				else if (!result) {
					state = 1;
					cout<<"Robot "<< robot << " crashes into robot " << board[dy][dx];
					break;
				}
				else if (result == -1) {
					state = 1;
					cout << "Robot " << robot << " crashes into the wall";
					break;
				}
			}
		}
		if (state) break;
	}

	if (!state) cout << "OK";

	return 0;
}